Exploratory Data Analysis of Rossmann Pharmaceuticals Sales Data


Imports

In [52]:
import numpy as np
import pandas as pd
import sys, os
import warnings
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px

In [53]:
sys.path.append(os.path.abspath(os.path.join('../scripts')))
from file_handler import FileHandler
from df_selector import *
from df_cleaner import *
from df_visualizer import *

In [54]:
warnings.filterwarnings('ignore')
pd.set_option('display.max_columns', 20)

Reading Data

In [55]:
# create a FileHandler object
file_handler = FileHandler()

In [56]:
# reading the store csv file
store_df = file_handler.read_csv("../data/rossmann-store-sales_data/store.csv")
store_df

Store StoreType Assortment  CompetitionDistance  \
0         1         c          a               1270.0   
1         2         a          a                570.0   
2         3         a          a              14130.0   
3         4         c          c                620.0   
4         5         a          a              29910.0   
...     ...       ...        ...                  ...   
1110   1111         a          a               1900.0   
1111   1112         c          c               1880.0   
1112   1113         a          c               9260.0   
1113   1114         a          c                870.0   
1114   1115         d          c               5350.0   

      CompetitionOpenSinceMonth  CompetitionOpenSinceYear  Promo2  \
0                           9.0                    2008.0       0   
1                          11.0                    2007.0       1   
2                          12.0                    2006.0       1   
3                           9.0                    2009.0       0   
4                           4.0                    2015.0       0   
...                         ...                       ...     ...   
1110                        6.0                    2014.0       1   
1111                        4.0                    2006.0       0   
1112                        NaN                       NaN       0   
1113                        NaN                       NaN       0   
1114                        NaN                       NaN       1   

      Promo2SinceWeek  Promo2SinceYear     PromoInterval  
0                 NaN              NaN               NaN  
1                13.0           2010.0   Jan,Apr,Jul,Oct  
2                14.0           2011.0   Jan,Apr,Jul,Oct  
3                 NaN              NaN               NaN  
4                 NaN              NaN               NaN  
...               ...              ...               ...  
1110             31.0           2013.0   Jan,Apr,Jul,Oct  
1111              NaN              NaN               NaN  
1112              NaN              NaN               NaN  
1113              NaN              NaN               NaN  
1114             22.0           2012.0  Mar,Jun,Sept,Dec  

[1115 rows x 10 columns]

In [57]:
# reading the sales training csv file
train_df = file_handler.read_csv("../data/rossmann-store-sales_data/train.csv")
train_df.head(10)

Store  DayOfWeek        Date  Sales  Customers  Open  Promo StateHoliday  \
0      1          5  2015-07-31   5263        555     1      1            0   
1      2          5  2015-07-31   6064        625     1      1            0   
2      3          5  2015-07-31   8314        821     1      1            0   
3      4          5  2015-07-31  13995       1498     1      1            0   
4      5          5  2015-07-31   4822        559     1      1            0   
5      6          5  2015-07-31   5651        589     1      1            0   
6      7          5  2015-07-31  15344       1414     1      1            0   
7      8          5  2015-07-31   8492        833     1      1            0   
8      9          5  2015-07-31   8565        687     1      1            0   
9     10          5  2015-07-31   7185        681     1      1            0   

   SchoolHoliday  
0              1  
1              1  
2              1  
3              1  
4              1  
5              1  
6              1  
7              1  
8              1  
9              1

General Statistics

store data

In [58]:
# rows and columns in the df
store_df.shape

(1115, 10)

In [59]:
store_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1115 entries, 0 to 1114
Data columns (total 10 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Store                      1115 non-null   int64  
 1   StoreType                  1115 non-null   object 
 2   Assortment                 1115 non-null   object 
 3   CompetitionDistance        1112 non-null   float64
 4   CompetitionOpenSinceMonth  761 non-null    float64
 5   CompetitionOpenSinceYear   761 non-null    float64
 6   Promo2                     1115 non-null   int64  
 7   Promo2SinceWeek            571 non-null    float64
 8   Promo2SinceYear            571 non-null    float64
 9   PromoInterval              571 non-null    object 
dtypes: float64(5), int64(2), object(3)
memory usage: 87.2+ KB


Train data

In [60]:
# rows and columns in the df
train_df.shape

(1017209, 9)

In [61]:
train_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1017209 entries, 0 to 1017208
Data columns (total 9 columns):
 #   Column         Non-Null Count    Dtype 
---  ------         --------------    ----- 
 0   Store          1017209 non-null  int64 
 1   DayOfWeek      1017209 non-null  int64 
 2   Date           1017209 non-null  object
 3   Sales          1017209 non-null  int64 
 4   Customers      1017209 non-null  int64 
 5   Open           1017209 non-null  int64 
 6   Promo          1017209 non-null  int64 
 7   StateHoliday   1017209 non-null  object
 8   SchoolHoliday  1017209 non-null  int64 
dtypes: int64(7), object(2)
memory usage: 69.8+ MB


Missing Values

Store data

In [62]:
percent_missing_values(store_df)

The dataset contains 21.01 % missing values.


In [63]:
missing_df = missing_values_table(store_df)

Your selected dataframe has 10 columns.
There are 6 columns that have missing values.


In [64]:
missing_df

Missing Values  % of Total Values    Dtype
Promo2SinceWeek                       544              48.79  float64
Promo2SinceYear                       544              48.79  float64
PromoInterval                         544              48.79   object
CompetitionOpenSinceMonth             354              31.75  float64
CompetitionOpenSinceYear              354              31.75  float64
CompetitionDistance                     3               0.27  float64

From the missing table we can recognize two groups with a pattern, the first 3 columns and the next 2 columns. Columns in each group have the same number of missing values. This indicates that if we investigate deeper we might get some insight.

Let's see the row number of rows with missing values in general then in each group.

In [65]:
# number of rows with missing values for the whole dataset
count_missing_rows(store_df)

750 rows(67.26%) contain atleast one missing value.


In [66]:
# number of rows with missing values for group 1 (Promo2SinceWeek, Promo2SinceYear, and PromoInterval)
count_missing_rows(store_df[['Promo2SinceWeek', 'Promo2SinceYear', 'PromoInterval']])

544 rows(48.79%) contain atleast one missing value.


In [67]:
# number of rows with missing values for group 2 (CompetitionOpenSinceMonth, and CompetitionOpenSinceYear)
count_missing_rows(store_df[['CompetitionOpenSinceMonth', 'CompetitionOpenSinceYear']])

354 rows(31.75%) contain atleast one missing value.


As guessed I have got an interesting insight. Each columuns in these groups have missing values in the same rows as their groupmates. These means the missing values in these groups appear at the same time as a cluster. From the column names and the data fields' description in the challange document we can understand that the values of the columns in each group share a common information. Let's investigate.

In [68]:
# dataframe containing the missing rows for columns in group 1
group1_df = store_df[store_df['Promo2SinceWeek'].isna()]
group1_df.head(10)

Store StoreType Assortment  CompetitionDistance  \
0       1         c          a               1270.0   
3       4         c          c                620.0   
4       5         a          a              29910.0   
5       6         a          a                310.0   
6       7         a          c              24000.0   
7       8         a          a               7520.0   
8       9         a          c               2030.0   
9      10         a          a               3160.0   
15     16         a          c               3270.0   
22     23         d          a               4060.0   

    CompetitionOpenSinceMonth  CompetitionOpenSinceYear  Promo2  \
0                         9.0                    2008.0       0   
3                         9.0                    2009.0       0   
4                         4.0                    2015.0       0   
5                        12.0                    2013.0       0   
6                         4.0                    2013.0       0   
7                        10.0                    2014.0       0   
8                         8.0                    2000.0       0   
9                         9.0                    2009.0       0   
15                        NaN                       NaN       0   
22                        8.0                    2005.0       0   

    Promo2SinceWeek  Promo2SinceYear PromoInterval  
0               NaN              NaN           NaN  
3               NaN              NaN           NaN  
4               NaN              NaN           NaN  
5               NaN              NaN           NaN  
6               NaN              NaN           NaN  
7               NaN              NaN           NaN  
8               NaN              NaN           NaN  
9               NaN              NaN           NaN  
15              NaN              NaN           NaN  
22              NaN              NaN           NaN

We can see that the Promo2 column is 0 for the first 10 rows. From intution we can infer that if there is no promotion the values for the columns in gropu 1(Promo2SinceWeek, Promo2SinceYear, and PromoInterval) should be null. Let's prove it.

In [69]:
# all unique values for each column
unique_values_df(group1_df)

Column  \
0                      Store   
1                  StoreType   
2                 Assortment   
3        CompetitionDistance   
4  CompetitionOpenSinceMonth   
5   CompetitionOpenSinceYear   
6                     Promo2   
7            Promo2SinceWeek   
8            Promo2SinceYear   
9              PromoInterval   

                                       Unique values  
0  [1, 758, 741, 742, 747, 751, 755, 757, 760, 4,...  
1                                       [a, d, c, b]  
2                                          [a, c, b]  
3  [250.0, 140.0, 50.0, 2640.0, 30.0, 420.0, 720....  
4  [9.0, 4.0, 11.0, 3.0, 12.0, 10.0, 7.0, 6.0, 2....  
5  [2013.0, 2012.0, 2014.0, 2009.0, 2010.0, 2005....  
6                                                [0]  
7                                                 []  
8                                                 []  
9                                                 []

In [70]:
group1_df['Promo2'].value_counts()

Promo2
0    544
Name: count, dtype: int64

All 544 values are 0. Now let's check the count of 0 in the Promo2 column for the whole store dataset.

In [71]:
store_df['Promo2'].value_counts()

Promo2
1    571
0    544
Name: count, dtype: int64

Yes, we are right. All rows with 0 value in their Promo2 columns have null values in the columns Promo2SinceWeek, Promo2SinceYear, and PromoInterval.


Now let's impute the missing values in the columns Promo2SinceWeek and Promo2SinceYear with 0 to convey the meaning of absence since year and week can't be 0. For PromoInterval we should dig deeper what kind of data it holds since its data type is object.

In [72]:
# check the exact data type of the object
pd.api.types.infer_dtype(store_df['PromoInterval'])

'string'

In [73]:
# all the unique values 
store_df['PromoInterval'].value_counts()

PromoInterval
Jan,Apr,Jul,Oct     335
Feb,May,Aug,Nov     130
Mar,Jun,Sept,Dec    106
Name: count, dtype: int64

For the column PromoInterval, we will impute it with '0,0,0,0' as the other values follow list-type format containing four months. So inorder to impute these columns we should first investigate if the values we will insert exist in each column as this will oppose our intention.

In [74]:
store_df[store_df['Promo2SinceWeek'] == 0].shape

(0, 10)

In [75]:
store_df[store_df['Promo2SinceYear'] == 0].shape

(0, 10)

In [76]:
store_df[store_df['PromoInterval'] == '0,0,0,0'].shape

(0, 10)

Since the values don't appear in these columns, let's fill the missing values by them.

In [77]:
fix_missing_value(store_df, ['Promo2SinceWeek', 'Promo2SinceYear'], 0)

544 missing values in the column Promo2SinceWeek have been replaced by 0.
544 missing values in the column Promo2SinceYear have been replaced by 0.


In [78]:
fix_missing_value(store_df, ['PromoInterval'], '0,0,0,0')

544 missing values in the column PromoInterval have been replaced by 0,0,0,0.


Now it's time for the second group (i.e CompetitionOpenSinceMonth and CompetitionOpenSinceYear).

In [79]:
# dataframe containing the missing rows for columns in group 1
group2_df = store_df[store_df['CompetitionOpenSinceMonth'].isna()]
group2_df.head(10)

Store StoreType Assortment  CompetitionDistance  \
11     12         a          c               1070.0   
12     13         d          a                310.0   
15     16         a          c               3270.0   
18     19         a          c               3240.0   
21     22         a          a               1040.0   
25     26         d          a               2300.0   
28     29         d          c               2170.0   
31     32         a          a               2910.0   
39     40         a          a                180.0   
40     41         d          c               1180.0   

    CompetitionOpenSinceMonth  CompetitionOpenSinceYear  Promo2  \
11                        NaN                       NaN       1   
12                        NaN                       NaN       1   
15                        NaN                       NaN       0   
18                        NaN                       NaN       1   
21                        NaN                       NaN       1   
25                        NaN                       NaN       0   
28                        NaN                       NaN       0   
31                        NaN                       NaN       1   
39                        NaN                       NaN       1   
40                        NaN                       NaN       1   

    Promo2SinceWeek  Promo2SinceYear     PromoInterval  
11             13.0           2010.0   Jan,Apr,Jul,Oct  
12             45.0           2009.0   Feb,May,Aug,Nov  
15              0.0              0.0           0,0,0,0  
18             22.0           2011.0  Mar,Jun,Sept,Dec  
21             22.0           2012.0   Jan,Apr,Jul,Oct  
25              0.0              0.0           0,0,0,0  
28              0.0              0.0           0,0,0,0  
31             45.0           2009.0   Feb,May,Aug,Nov  
39             45.0           2009.0   Feb,May,Aug,Nov  
40             31.0           2013.0   Jan,Apr,Jul,Oct

In [80]:
# all unique values for each column in the dataframe
unique_values_df(group2_df)

Column  \
0                      Store   
1                  StoreType   
2                 Assortment   
3        CompetitionDistance   
4  CompetitionOpenSinceMonth   
5   CompetitionOpenSinceYear   
6                     Promo2   
7            Promo2SinceWeek   
8            Promo2SinceYear   
9              PromoInterval   

                                       Unique values  
0  [12, 853, 767, 766, 764, 762, 757, 756, 742, 7...  
1                                       [a, d, c, b]  
2                                          [c, a, b]  
3  [250.0, 2410.0, 50.0, 210.0, 840.0, 220.0, 150...  
4                                                 []  
5                                                 []  
6                                             [1, 0]  
7  [0.0, 40.0, 14.0, 45.0, 31.0, 22.0, 5.0, 1.0, ...  
8  [0.0, 2013.0, 2011.0, 2009.0, 2012.0, 2014.0, ...  
9  [0,0,0,0, Jan,Apr,Jul,Oct, Feb,May,Aug,Nov, Ma...

I can't see any pattern with the other columns. But from the column names and the data fields' description in the challange document I can understand that if CompetitionOpenSinceMonth and CompetitionOpenSinceYear are null it means there was already a competition when the store was opened. So we need impute the missing values in a meaningful way. Before doing that we will hold this meaning of an already existing competition in a separate column named CompetitionBeforeStoreOpened so that we can make use of this information in our analysis and prediction. This column will hold values of 0 and 1, 0 indicating the absence of competition at the time the store was opened and 1 indicating the presence of competition at the time the store was opened.

In [81]:
# deriving new column
store_df['CompetitionBeforeStoreOpened'] = store_df['CompetitionOpenSinceYear'].apply(lambda x: 1 if np.isnan(x) else 0)
store_df

Store StoreType Assortment  CompetitionDistance  \
0         1         c          a               1270.0   
1         2         a          a                570.0   
2         3         a          a              14130.0   
3         4         c          c                620.0   
4         5         a          a              29910.0   
...     ...       ...        ...                  ...   
1110   1111         a          a               1900.0   
1111   1112         c          c               1880.0   
1112   1113         a          c               9260.0   
1113   1114         a          c                870.0   
1114   1115         d          c               5350.0   

      CompetitionOpenSinceMonth  CompetitionOpenSinceYear  Promo2  \
0                           9.0                    2008.0       0   
1                          11.0                    2007.0       1   
2                          12.0                    2006.0       1   
3                           9.0                    2009.0       0   
4                           4.0                    2015.0       0   
...                         ...                       ...     ...   
1110                        6.0                    2014.0       1   
1111                        4.0                    2006.0       0   
1112                        NaN                       NaN       0   
1113                        NaN                       NaN       0   
1114                        NaN                       NaN       1   

      Promo2SinceWeek  Promo2SinceYear     PromoInterval  \
0                 0.0              0.0           0,0,0,0   
1                13.0           2010.0   Jan,Apr,Jul,Oct   
2                14.0           2011.0   Jan,Apr,Jul,Oct   
3                 0.0              0.0           0,0,0,0   
4                 0.0              0.0           0,0,0,0   
...               ...              ...               ...   
1110             31.0           2013.0   Jan,Apr,Jul,Oct   
1111              0.0              0.0           0,0,0,0   
1112              0.0              0.0           0,0,0,0   
1113              0.0              0.0           0,0,0,0   
1114             22.0           2012.0  Mar,Jun,Sept,Dec   

      CompetitionBeforeStoreOpened  
0                                0  
1                                0  
2                                0  
3                                0  
4                                0  
...                            ...  
1110                             0  
1111                             0  
1112                             1  
1113                             1  
1114                             1  

[1115 rows x 11 columns]

Now we can impute the columns CompetitionOpenSinceMonth and CompetitionOpenSinceYear. So we will impute missing values in the column CompetitionOpenSinceYear and CompetitionOpenSinceMonth with its minimun value and 1 respectively. This kind of imputing assumes that the stores with missing values in these columns were opened after the first month of the minimum year value in the column CompetitionOpenSinceYear.

Let's find the minimum year value for CompetitionOpenSinceYear and impute the missing values.

In [82]:
min_year = store_df['CompetitionOpenSinceYear'].min()
min_year

1900.0

In [83]:
# impute the column CompetitionOpenSinceYear with 1900 and the column CompetitionOpenSinceMonth with 1
fix_missing_value(store_df, ['CompetitionOpenSinceYear'], min_year)
fix_missing_value(store_df, ['CompetitionOpenSinceMonth'], 1)

354 missing values in the column CompetitionOpenSinceYear have been replaced by 1900.0.
354 missing values in the column CompetitionOpenSinceMonth have been replaced by 1.


In [84]:
store_df

Store StoreType Assortment  CompetitionDistance  \
0         1         c          a               1270.0   
1         2         a          a                570.0   
2         3         a          a              14130.0   
3         4         c          c                620.0   
4         5         a          a              29910.0   
...     ...       ...        ...                  ...   
1110   1111         a          a               1900.0   
1111   1112         c          c               1880.0   
1112   1113         a          c               9260.0   
1113   1114         a          c                870.0   
1114   1115         d          c               5350.0   

      CompetitionOpenSinceMonth  CompetitionOpenSinceYear  Promo2  \
0                           9.0                    2008.0       0   
1                          11.0                    2007.0       1   
2                          12.0                    2006.0       1   
3                           9.0                    2009.0       0   
4                           4.0                    2015.0       0   
...                         ...                       ...     ...   
1110                        6.0                    2014.0       1   
1111                        4.0                    2006.0       0   
1112                        1.0                    1900.0       0   
1113                        1.0                    1900.0       0   
1114                        1.0                    1900.0       1   

      Promo2SinceWeek  Promo2SinceYear     PromoInterval  \
0                 0.0              0.0           0,0,0,0   
1                13.0           2010.0   Jan,Apr,Jul,Oct   
2                14.0           2011.0   Jan,Apr,Jul,Oct   
3                 0.0              0.0           0,0,0,0   
4                 0.0              0.0           0,0,0,0   
...               ...              ...               ...   
1110             31.0           2013.0   Jan,Apr,Jul,Oct   
1111              0.0              0.0           0,0,0,0   
1112              0.0              0.0           0,0,0,0   
1113              0.0              0.0           0,0,0,0   
1114             22.0           2012.0  Mar,Jun,Sept,Dec   

      CompetitionBeforeStoreOpened  
0                                0  
1                                0  
2                                0  
3                                0  
4                                0  
...                            ...  
1110                             0  
1111                             0  
1112                             1  
1113                             1  
1114                             1  

[1115 rows x 11 columns]

In [85]:
temp_df = store_df[store_df['CompetitionDistance'].isna()]
temp_df

Store StoreType Assortment  CompetitionDistance  \
290    291         d          a                  NaN   
621    622         a          c                  NaN   
878    879         d          a                  NaN   

     CompetitionOpenSinceMonth  CompetitionOpenSinceYear  Promo2  \
290                        1.0                    1900.0       0   
621                        1.0                    1900.0       0   
878                        1.0                    1900.0       1   

     Promo2SinceWeek  Promo2SinceYear    PromoInterval  \
290              0.0              0.0          0,0,0,0   
621              0.0              0.0          0,0,0,0   
878              5.0           2013.0  Feb,May,Aug,Nov   

     CompetitionBeforeStoreOpened  
290                             1  
621                             1  
878                             1

In [86]:
unique_values_df(temp_df)

Column               Unique values
0                          Store             [291, 622, 879]
1                      StoreType                      [d, a]
2                     Assortment                      [a, c]
3            CompetitionDistance                          []
4      CompetitionOpenSinceMonth                       [1.0]
5       CompetitionOpenSinceYear                    [1900.0]
6                         Promo2                      [0, 1]
7                Promo2SinceWeek                  [0.0, 5.0]
8                Promo2SinceYear               [0.0, 2013.0]
9                  PromoInterval  [0,0,0,0, Feb,May,Aug,Nov]
10  CompetitionBeforeStoreOpened                         [1]

All rows with missing values in the column CompetitionDistance has a value of 1 in the column CompetitionBeforeStoreOpened. But this doesn't give us much information why the CompetitionDistance is null. But from common sense I guess this is because the competitions for these stores are very far away and have almost no impact on these stores that they weren't measured when collecting the data. For this reason, I will impute this values with the maximum competition distance.

In [87]:
max_dist = store_df['CompetitionDistance'].max()
max_dist

75860.0

In [88]:
fix_missing_value(store_df, ['CompetitionDistance'], max_dist)

3 missing values in the column CompetitionDistance have been replaced by 75860.0.


In [89]:
# final check for missing values
percent_missing_values(store_df)

The dataset contains 0.0 % missing values.


Train data

In [90]:
percent_missing_values(train_df)

The dataset contains 0.0 % missing values.


Data Types

store data

In [91]:
store_df.dtypes

Store                             int64
StoreType                        object
Assortment                       object
CompetitionDistance             float64
CompetitionOpenSinceMonth       float64
CompetitionOpenSinceYear        float64
Promo2                            int64
Promo2SinceWeek                 float64
Promo2SinceYear                 float64
PromoInterval                    object
CompetitionBeforeStoreOpened      int64
dtype: object

In [92]:
store_df.head()

Store StoreType Assortment  CompetitionDistance  CompetitionOpenSinceMonth  \
0      1         c          a               1270.0                        9.0   
1      2         a          a                570.0                       11.0   
2      3         a          a              14130.0                       12.0   
3      4         c          c                620.0                        9.0   
4      5         a          a              29910.0                        4.0   

   CompetitionOpenSinceYear  Promo2  Promo2SinceWeek  Promo2SinceYear  \
0                    2008.0       0              0.0              0.0   
1                    2007.0       1             13.0           2010.0   
2                    2006.0       1             14.0           2011.0   
3                    2009.0       0              0.0              0.0   
4                    2015.0       0              0.0              0.0   

     PromoInterval  CompetitionBeforeStoreOpened  
0          0,0,0,0                             0  
1  Jan,Apr,Jul,Oct                             0  
2  Jan,Apr,Jul,Oct                             0  
3          0,0,0,0                             0  
4          0,0,0,0                             0

As we can see from the dataframe above StoreType, Assortment and PromoInterval hold string values. Note that even though PromoInterval holds list values, the are enclosed in string literals. We will convert the data types of these three columns to string. Converting PromoInterval to string is convenient for further processing elements in each value.

In [93]:
# get the columns with object data type
string_columns = store_df.select_dtypes(include='object').columns.tolist()
string_columns

['StoreType', 'Assortment', 'PromoInterval']

In [94]:
convert_to_string(store_df, string_columns)

We also know that the columns CompetitionOpenSinceMonth, CompetitionOpenSinceYear, Promo2SinceWeek, and Promo2SinceYear contain week, month and year information which can't be float. So we will change the data type of these columns to int64.

In [95]:
convert_to_int(store_df, ['CompetitionOpenSinceMonth',  'CompetitionOpenSinceYear',
        'Promo2SinceWeek', 'Promo2SinceYear'])

In [96]:
store_df.dtypes

Store                                    int64
StoreType                       string[python]
Assortment                      string[python]
CompetitionDistance                    float64
CompetitionOpenSinceMonth                int64
CompetitionOpenSinceYear                 int64
Promo2                                   int64
Promo2SinceWeek                          int64
Promo2SinceYear                          int64
PromoInterval                   string[python]
CompetitionBeforeStoreOpened             int64
dtype: object

Train data

In [97]:
show_cols_mixed_dtypes(train_df)

         Column      Data type
0  StateHoliday  mixed-integer


So StateHoliday needs more investigation of the datatypes it contain. Let's look at the unique values.

In [98]:
train_df['StateHoliday'].value_counts()

StateHoliday
0    855087
0    131072
a     20260
b      6690
c      4100
Name: count, dtype: int64

In [99]:
train_df['StateHoliday'].value_counts().index

Index(['0', 0, 'a', 'b', 'c'], dtype='object', name='StateHoliday')

Here 0 is represented as an integer and a string. Since there are other string values we will convert the StateHoliday column into string.

In [100]:
convert_to_string(train_df, ['StateHoliday'])

train_df.dtypes

Store                     int64
DayOfWeek                 int64
Date                     object
Sales                     int64
Customers                 int64
Open                      int64
Promo                     int64
StateHoliday     string[python]
SchoolHoliday             int64
dtype: object

In [101]:
train_df.head()

Store  DayOfWeek        Date  Sales  Customers  Open  Promo StateHoliday  \
0      1          5  2015-07-31   5263        555     1      1            0   
1      2          5  2015-07-31   6064        625     1      1            0   
2      3          5  2015-07-31   8314        821     1      1            0   
3      4          5  2015-07-31  13995       1498     1      1            0   
4      5          5  2015-07-31   4822        559     1      1            0   

   SchoolHoliday  
0              1  
1              1  
2              1  
3              1  
4              1

By observing the columns we can realize that the Date column actually contain datetime values even though its data type is object. So we will convert it to datetime data type.

In [102]:
convert_to_datetime(train_df, ['Date'])

train_df.dtypes

Store                     int64
DayOfWeek                 int64
Date             datetime64[ns]
Sales                     int64
Customers                 int64
Open                      int64
Promo                     int64
StateHoliday     string[python]
SchoolHoliday             int64
dtype: object

Duplicates